test/performance/odp_ipsecfwd.c

```diff
@@ -949,25 +994,203 @@ static odp_bool_t create_sa_dest_queues(odp_ipsec_capability_t *ipsec_capa,
 	return true;
 }
 
-static void create_sa_entry(uint32_t dir, uint32_t spi, const char *src_ip_str,
-			    const char *dst_ip_str, int cipher_idx, uint8_t *cipher_key,
-			    uint8_t *cipher_key_extra, int auth_idx, uint8_t *auth_key,
-			    uint8_t *auth_key_extra, uint32_t icv_len, uint32_t ar_ws,
-			    uint32_t max_num_sa, prog_config_t *config)
+static void parse_crypto(config_setting_t *cfg, sa_config_t *config)
 {
-	uint32_t src_ip, dst_ip;
-	odp_ipsec_sa_param_t sa_param;
-	odp_ipsec_crypto_param_t crypto_param;
-	odp_ipsec_sa_t sa;
+	int val;
+	const char *val_str;
+	config_setting_t *cs = config_setting_lookup(cfg, "crypto");
 
-	if (config->num_sas == max_num_sa) {
-		ODPH_ERR("Maximum number of SAs parsed (%u), ignoring rest\n", max_num_sa);
+	if (cs == NULL)
+		return;
+
+	if (config_setting_lookup_int(cs, "cipher_alg", &val) == CONFIG_TRUE)
+		config->sa_param.crypto.cipher_alg = val;
+
+	if (config_setting_lookup_string(cs, "cipher_key", &val_str) == CONFIG_TRUE) {
+		strcpy(config->cipher_key, val_str);
+		config->sa_param.crypto.cipher_key.data = (uint8_t *)config->cipher_key;
+		config->sa_param.crypto.cipher_key.length =
+			strlen((const char *)config->cipher_key);
+	}
+
+	if (config_setting_lookup_string(cs, "cipher_key_extra", &val_str) == CONFIG_TRUE) {
+		strcpy(config->cipher_key_extra, val_str);
+		config->sa_param.crypto.cipher_key_extra.data =
+			(uint8_t *)config->cipher_key_extra;
+		config->sa_param.crypto.cipher_key_extra.length =
+			strlen((const char *)config->cipher_key_extra);
+	}
+
+	if (config_setting_lookup_int(cs, "auth_alg", &val) == CONFIG_TRUE)
+		config->sa_param.crypto.auth_alg = val;
+
+	if (config_setting_lookup_string(cs, "auth_key", &val_str) == CONFIG_TRUE) {
+		strcpy(config->auth_key, val_str);
+		config->sa_param.crypto.auth_key.data = (uint8_t *)config->auth_key;
+		config->sa_param.crypto.auth_key.length = strlen((const char *)config->auth_key);
+	}
+
+	if (config_setting_lookup_string(cs, "auth_key_extra", &val_str) == CONFIG_TRUE) {
+		strcpy(config->auth_key_extra, val_str);
+		config->sa_param.crypto.auth_key_extra.data = (uint8_t *)config->auth_key_extra;
+		config->sa_param.crypto.auth_key_extra.length =
+			strlen((const char *)config->auth_key_extra);
+	}
+
+	(void)config_setting_lookup_int(cs, "icv_len", (int *)&config->sa_param.crypto.icv_len);
```

Comment:
This could follow the same pattern as the other settings so that the casting (and the assumption that icv_len is the same size as int) would not be needed. And the same for mtu and spi parameters. With that, Reviewed-by: Janne Peltonen <janne.peltonen@nokia.com>

Reply:
V9.